docs/proposals/20210305-sidecarset-hotupgrade.md

```diff
@@ -0,0 +1,276 @@
+State migration is a very important step in the hot upgrade process, so i will describe some of the key points below:
+
+**1. Consultation Mechanism**
+
+The hot migration process starts with a negotiation between the two sidecar containers to decide the migration direction (sidecar-1 -> sidecar-2 or sidecar-2 -> sidecar-1).
+Since state migration is always from old -> new version. For this case OpenKruise injects the environment variable **SIDECARSET_VERSION** into the sidecar container by default.
+The value of SIDECARSET_VERSION is the resourceVersion of the corresponding SidecarSet, so the larger one is the newer version compared with the two SIDECARSET_VERSION. As shown below, envoy-2 is the new version of sidecar:
+```
+apiVersion: v1
+kind: Pod
+spec:
+  containers:
+  - name: envoy-1
+    env:
+    - name: SIDECARSET_VERSION
+      value: '1097936677'
+  - name: envoy-2
+    env:
+    - name: SIDECARSET_VERSION
+      value: '1098297016'
+```
+Although the version of sidecar has been saved in the environment SIDECARSET_VERSION, it still needs to be negotiated between two sidecars to determine, and recommend to use shared Linux IPC, such as: TCP communication, shared memory, UDS, etc.
+The specific business can decide by itself according to its own technology stack.
+
+**NOTE: The next release will include environment SIDECARSET_VERSION_ALT( the other sidecar version ) in the container, so negotiation is no longer required**
+
+**2. State Migration**
+
+State migration is essentially the migration of stateful resources from the old version to the new version without interrupting service to the outside world. For example, for envoy, it is the migration of fd from the old sidecar to the new sidecar. fd can be passed between processes via Unix Domain Socket, but simply passing fd is still far from the goal of hot upgrade.
+We can divide fd into two types: ListenFd and ConnectionFd, ListenFd is easier to migrate due to its stateless property, while ConnectionFd is not easy to complete the migration, the recommended migration solution is to migrate ListenFd, ConnectionFd "gracefully offline" and re-establish the connection to the new service process at the next request.
+
+**3. Exception Handling**
+
+Any abnormalities in the hot upgrade process will result in Pod ready=false, that is, the hot upgrade is lossy, so we only focus on whether the hot upgrade can end normally:
+- Old version of the container abnormal exit: kubelet will restart the exit container, PostStartHook script needs to be compatible with such scenarios, return to normal, the hot upgrade process will proceed normally
+- PostStartHook failure: kubelet will restart the new version of the container and execute the script PostStartHook failure until it succeeds
+
+#### Other Notes
+1. Because for the hot upgrade sidecar will be injected into two sidecar containers (resources.requests will be double), so recommended that resources.requests is not set or set to half
```

Comment:
For a single  hot upgrade sidecar in the sidecarset,  two sidecar containers  will be injected (resources.requests will double), so we recommend to leave the resources.requests unset, or set to half of the original  resources.requests